grovedb/src/replication.rs

```diff
@@ -1,990 +1,430 @@
-            }
-        }
-
-        if self.cache.is_none() {
-            let current_merk = self
-                .grove_db
-                .open_non_transactional_merk_at_path(
-                    path_iter.clone().collect::<Vec<_>>().as_slice().into(),
-                    None,
-                )
-                .unwrap()?;
-
-            if current_merk.root_key().is_none() {
-                return Ok(Vec::new());
-            }
-
-            self.cache = Some(SubtreeChunkProducerCache {
-                current_merk_path: path_iter.map(|p| p.to_vec()).collect(),
-                current_merk,
-                current_chunk_producer: None,
-            });
-            let cache = self.cache.as_mut().expect("exists at this point");
-            cache.current_chunk_producer = Some(
-                grovedb_merk::ChunkProducer::new(&cache.current_merk)
-                    .map_err(|e| Error::CorruptedData(e.to_string()))?,
+impl fmt::Debug for SubtreesMetadata {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for (prefix, metadata) in self.data.iter() {
+            let metadata_path = &metadata.0;
+            let metadata_path_str = util_path_to_string(metadata_path);
+            writeln!(
+                f,
+                " prefix:{:?} -> path:{:?}\n",
+                hex::encode(prefix),
+                metadata_path_str
             );
         }
-
-        self.cache
-            .as_mut()
-            .expect("must exist at this point")
-            .current_chunk_producer
-            .as_mut()
-            .expect("must exist at this point")
-            .chunk(index)
-            .map_err(|e| Error::CorruptedData(e.to_string()))
+        Ok(())
     }
 }
 
-// TODO: make generic over storage_cost context
-type MerkRestorer<'db> = grovedb_merk::Restorer<PrefixedRocksDbImmediateStorageContext<'db>>;
-
-type Path = Vec<Vec<u8>>;
-
-/// Structure to drive GroveDb restore process.
-pub struct Restorer<'db> {
-    current_merk_restorer: Option<MerkRestorer<'db>>,
-    current_merk_chunk_index: usize,
-    current_merk_path: Path,
-    queue: VecDeque<(Path, Vec<u8>, Hash, TreeFeatureType)>,
-    grove_db: &'db GroveDb,
-    tx: &'db Transaction<'db>,
+// Converts a path into a human-readable string (for debuting)
+pub fn util_path_to_string(path: &[Vec<u8>]) -> Vec<String> {
+    let mut subtree_path_str: Vec<String> = vec![];
+    for subtree in path {
+        let string = std::str::from_utf8(subtree).unwrap();
+        subtree_path_str.push(string.parse().unwrap());
+    }
+    subtree_path_str
 }
 
-/// Indicates what next piece of information `Restorer` expects or wraps a
-/// successful result.
-#[derive(Debug)]
-pub enum RestorerResponse {
-    AwaitNextChunk { path: Vec<Vec<u8>>, index: usize },
-    Ready,
-}
+// Splits the given global chunk id into [SUBTREE_PREFIX:CHUNK_ID]
+pub fn util_split_global_chunk_id(
+    global_chunk_id: &[u8],
+) -> Result<(crate::SubtreePrefix, String), Error> {
```

Comment:
why is this a string and not a Vec<u8>?

Reply:
@QuantumExplorer Existing work on merk chunking makes use of string for chunk identification.
This PR, uses Vec<u8> for chunk identification in all APIs for ABCI, but continue to use internally string in merk chunking.
I agree that this should be changed to Vec<u8> everywhere but I propose to make this change on a separate PR.

Reply:
Okay in a future PR then